# Laboratorio de regresión logística

|                |   |
:----------------|---|
| **Nombre**     | GIBRÁN LEONARDO CHÁVEZ GONZÁLEZ  |
| **Fecha**      | 16/03/26  |
| **Expediente** | 757366  | 

In machine learning, Support Vector Machines (SVM) are supervised learning models with associated learning algorithms that analyze data used for classification and regression analysis. It is mostly used in classification problems. In this algorithm, each data item is plotted as a point in p-dimensional space (where p is the number of features), with the value of each feature being the value of a particular coordinate. Then, classification is performed by finding the hyper-plane that best differentiates the two classes (or more if we have a multi class problem):

$$ f(x) = w^T \varphi(x) + b $$

where $\varphi: X \rightarrow F $ is a function that makes each input point $x$ correspond to a point in F, where F is a Hilbert space.

In addition to performing linear classification, SVMs can efficiently perform a non-linear classification, implicitly mapping their inputs into high-dimensional feature spaces (more specifically using the kernel trick, like the RBF funcion). 

[1]

OLS utilizes the squared residuals to fit the parameters. Large residuals caused by outliers may worsen the accuracy significantly.

Support Vectors use piecewise linear functions to counter this, in which a hyperparameter  $\epsilon$ called the margin lets errors that are less or equal to it be 0, and error larger than it be $e - \epsilon$. 

The problem to solve is:

\begin{split}
        \min_{w, b, \xi, \xi^*} \mathcal{P}_\epsilon(w, b, \xi) &= \frac{1}{2} w^T w + c \sum_{k=1}^{N} \xi_k \\
        \text{s.t. } & y_k [w^T \varphi(x_k) - b] \geq 1- \xi_k,\ \ k = 1, ..., N \\
        & \xi_k \geq 0,\ \ k = 1, ..., N
\end{split}


The most important question that arises when using a SVM is how to choose the correct hyperplane. Consider the following scenarios:

### Scenario 1

In this scenario there are three hyperplanes called A, B, and C. Now, the problem is to identify the hyperplane which best differentiates the stars and the circles.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_21-2.png" alt="what image shows"></center>

In this case, hyperplane B separates the stars and the circle betters, hence it is the correct hyperplane.


### Scenario 2

Now take another scenario where all three hyperplanes are segregating classes well. The question that arises is how to choose the best hyperplane in this situation.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_4-2.png" alt="what image shows"></center>

In such scenarios, we calculate the margin (which is the distance between nearest data point and the hyperplane). The hyperplane with the largest margin will be considered as the correct hyperplane to classify the dataset.

Here C has the largest margin. Hence, it is considered as the best hyperplane.


### Kernels
Knowing 
$$ w = \sum_{k=1}^{N} \alpha_k y_k \varphi(x_k) $$

And
$$ y_{pred} = w^T \varphi(x) + b $$

Then 
$$ y_{pred} = (\sum_{k=1}^{N} \alpha_k y_k \varphi(x_k))^T \varphi(x) + b $$

Where $\varphi$ is a function that makes each input in $x$ correspond to a point in $F$ (a Hilbert space). This can be seen as processing and transforming the input featuers to keep the model's convexity. [2]

This also allows us to transform the inputs into another space where they might be more easily classified.

<center><img src=https://miro.medium.com/max/838/1*gXvhD4IomaC9Jb37tzDUVg.png alt="what image shows"></center>

## ROC and AUC

A ROC (Receiver Operating Characteristic) is a graph that shows how the classification model performs at the classification thresholds. 

ROC curves typically feature true positive rate on the Y axis, and false positive rate on the X axis. This means that the top left corner of the plot is the “ideal” point - a false positive rate of zero, and a true positive rate of one. This is not very realistic, but it does mean that a larger area under the curve (AUC) is usually better. [3]

True Positive Rate is a synonym for Recall and defined as:
$$ TPR = \frac{TP}{TP + FN} $$

False Positive Rate is a synonym for Specificity and defined as:

$$ FPR = \frac{FP}{FP + TN} $$

ROC curves are typically used in binary classification to study the output of a classifier. In order to extend ROC curve and ROC area to multi-label classification, it is necessary to binarize the output. One ROC curve can be drawn per label, but one can also draw a ROC curve by considering each element of the label indicator matrix as a binary prediction (micro-averaging).

E.g. If you lower a classification threshold, more items would be classified as positive, increasing False Positives and True Positives.

AUC stands for Area under the ROC.

## Ejercicio 1

- Utiliza el dataset `Iris`, modela con SVC y haz Cross-Validation de diferentes kernels ('linear', 'poly', 'rbf', 'sigmoid').
- Modela con LogisticRegression.
- El método de Cross-Validation es K-Folds con $k=10$.
- Utiliza el AUC como métrico de Cross-Validation.
- Compara resultados.

In [33]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.model_selection import cross_val_score

In [34]:
# Carga del dataset Iris
from sklearn.datasets import load_iris

iris = load_iris()

X_iris = iris.data
y_iris = iris.target

In [35]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [36]:
kernels = ['linear', 'poly', 'rbf', 'sigmoid']

In [37]:
resultados_iris = {}

for kernel_ in kernels:
    model = SVC(probability=True, kernel=kernel_)

    pipeline = Pipeline(steps=[
    # Etiqueta, proceso
    ('model', model)])

    pipeline.fit(X_iris, y_iris)

    cross_val = cross_val_score(pipeline, X_iris, y_iris, scoring='roc_auc_ovr', cv=10) # AUC: Area Under the Curve

    resultados_iris[f'SVC_{kernel_}'] = cross_val

    print(f'Kernel: {kernel_}, Parámetros: {pipeline.get_params()}')
    print(f'Kernel: {kernel_}, Cross Validation: {cross_val}\n\n\n')


Kernel: linear, Parámetros: {'memory': None, 'steps': [('model', SVC(kernel='linear', probability=True))], 'transform_input': None, 'verbose': False, 'model': SVC(kernel='linear', probability=True), 'model__C': 1.0, 'model__break_ties': False, 'model__cache_size': 200, 'model__class_weight': None, 'model__coef0': 0.0, 'model__decision_function_shape': 'ovr', 'model__degree': 3, 'model__gamma': 'scale', 'model__kernel': 'linear', 'model__max_iter': -1, 'model__probability': True, 'model__random_state': None, 'model__shrinking': True, 'model__tol': 0.001, 'model__verbose': False}
Kernel: linear, Cross Validation: [1.         1.         1.         1.         1.         1.
 0.98666667 1.         1.         1.        ]



Kernel: poly, Parámetros: {'memory': None, 'steps': [('model', SVC(kernel='poly', probability=True))], 'transform_input': None, 'verbose': False, 'model': SVC(kernel='poly', probability=True), 'model__C': 1.0, 'model__break_ties': False, 'model__cache_size': 200, 'model__c

In [38]:
model = LogisticRegression()

pipeline = Pipeline(steps=[
# Etiqueta, proceso
('model', model)])

pipeline.fit(X_iris, y_iris)

cross_val_iris = cross_val_score(pipeline, X_iris, y_iris, scoring='roc_auc_ovr', cv=10) # AUC: Area Under the Curve

resultados_iris['LogReg'] = cross_val_iris

print(f'Cross Validation: {cross_val_iris}\n\n\n')

Cross Validation: [1.         1.         1.         1.         1.         1.
 0.98666667 1.         1.         1.        ]





C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\linear_model\_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\linear_model\_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the 

In [39]:
resultados_iris

{'SVC_linear': array([1.        , 1.        , 1.        , 1.        , 1.        ,
        1.        , 0.98666667, 1.        , 1.        , 1.        ]),
 'SVC_poly': array([1.        , 1.        , 1.        , 1.        , 1.        ,
        1.        , 0.98666667, 1.        , 1.        , 1.        ]),
 'SVC_rbf': array([1.        , 1.        , 1.        , 1.        , 1.        ,
        0.98666667, 0.98666667, 1.        , 1.        , 1.        ]),
 'SVC_sigmoid': array([0.79333333, 0.71333333, 0.68      , 0.64666667, 0.83333333,
        0.8       , 0.66      , 0.76      , 0.72      , 0.74      ]),
 'LogReg': array([1.        , 1.        , 1.        , 1.        , 1.        ,
        1.        , 0.98666667, 1.        , 1.        , 1.        ])}

In [40]:
df_comparacion_iris = pd.DataFrame({
    'Modelo': resultados_iris.keys(),
    'Media': [np.mean(values) for values in resultados_iris.values()],
    'Desviación Estándar': [np.std(values) for values in resultados_iris.values()]
})

df_comparacion_iris

,Modelo,Media,Desviación Estándar
0,SVC_linear,0.998667,0.004000
1,SVC_poly,0.998667,0.004000
2,SVC_rbf,0.997333,0.005333
3,SVC_sigmoid,0.734667,0.059240
4,LogReg,0.998667,0.004000


De acuerdo con la media y la desviación estándar del AUC obtenidas, los modelos SVC con kernel linear, poly y LogisticRegression presentaron el mejor desempeño, con una media de 0.998667 y baja variabilidad. El modelo SVC con kernel RBF también mostró un buen rendimiento, aunque inferior. En contraste, SVC con kernel sigmoid obtuvo la media más baja y la mayor desviación estándar, lo que indica un peor desempeño y menor estabilidad. 

## Ejercicio 2
- Repite el ejercicio 1 con el dataset `Default`. Utiliza `default` como target.

In [41]:
df = pd.read_csv('Default.csv')
num_cols = ['balance', 'income']
cat_cols = ['student']
target_col = ['default']
X_default = df[num_cols + cat_cols]
y_default = df[target_col]

In [42]:
num_transformer = StandardScaler()
cat_transformer = OneHotEncoder()

preprocessor = ColumnTransformer(
    transformers=[('num', num_transformer, num_cols), # Etiquetra, tipo de transformación, a qué se lo aplicas
                 ('cat', cat_transformer, cat_cols)
    ]
)

In [43]:
resultados_default = {}

for kernel_ in kernels:
    model = SVC(probability=True, kernel=kernel_)

    pipeline = Pipeline(steps=[
    # Etiqueta, proceso
    ('preprocessor', preprocessor),
    ('model', model)])

    pipeline.fit(X_default, y_default)

    cross_val = cross_val_score(pipeline, X_default, y_default, scoring='roc_auc_ovr', cv=10) # AUC: Area Under the Curve

    resultados_default[f'SVC_{kernel_}'] = cross_val

    print(f'Kernel: {kernel_}, Cross Validation: {cross_val}\n\n\n')

C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the sha

Kernel: linear, Cross Validation: [0.94017737 0.85468961 0.96220739 0.06207891 0.93528877 0.93604086
 0.95080066 0.94449519 0.90587322 0.89836804]





C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the sha

Kernel: poly, Cross Validation: [0.80906271 0.79583843 0.86070634 0.8245746  0.78975902 0.86431011
 0.8658143  0.82855316 0.83372914 0.79670564]





C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the sha

Kernel: rbf, Cross Validation: [0.80332801 0.7928614  0.86070634 0.87205039 0.80555294 0.87712701
 0.84419166 0.84131044 0.85692973 0.81320789]





C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the sha

Kernel: sigmoid, Cross Validation: [0.70543073 0.60919432 0.82140955 0.76625615 0.74811194 0.7612422
 0.68302466 0.70509073 0.73502009 0.66222141]





In [44]:
model = LogisticRegression()

pipeline = Pipeline(steps=[
# Etiqueta, proceso
('preprocessor', preprocessor),
('model', model)])

pipeline.fit(X_default, y_default)

cross_val_default = cross_val_score(pipeline, X_default, y_default, scoring='roc_auc_ovr', cv=10) # AUC: Area Under the Curve

resultados_default['LogReg'] = cross_val_default

print(f'Cross Validation: {cross_val_default}\n\n\n')

C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the sha

Cross Validation: [0.94506596 0.96023315 0.96615587 0.95202281 0.94362446 0.93660493
 0.97207859 0.95253319 0.94595664 0.92443064]





C:\Users\gibra\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [45]:
resultados_default

{'SVC_linear': array([0.94017737, 0.85468961, 0.96220739, 0.06207891, 0.93528877,
        0.93604086, 0.95080066, 0.94449519, 0.90587322, 0.89836804]),
 'SVC_poly': array([0.80906271, 0.79583843, 0.86070634, 0.8245746 , 0.78975902,
        0.86431011, 0.8658143 , 0.82855316, 0.83372914, 0.79670564]),
 'SVC_rbf': array([0.80332801, 0.7928614 , 0.86070634, 0.87205039, 0.80555294,
        0.87712701, 0.84419166, 0.84131044, 0.85692973, 0.81320789]),
 'SVC_sigmoid': array([0.70543073, 0.60919432, 0.82140955, 0.76625615, 0.74811194,
        0.7612422 , 0.68302466, 0.70509073, 0.73502009, 0.66222141]),
 'LogReg': array([0.94506596, 0.96023315, 0.96615587, 0.95202281, 0.94362446,
        0.93660493, 0.97207859, 0.95253319, 0.94595664, 0.92443064])}

In [46]:
df_comparacion_default = pd.DataFrame({
    'Modelo': resultados_default.keys(),
    'Media': [np.mean(values) for values in resultados_default.values()],
    'Desviación Estándar': [np.std(values) for values in resultados_default.values()]
})

df_comparacion_default

,Modelo,Media,Desviación Estándar
0,SVC_linear,0.839002,0.260688
1,SVC_poly,0.826905,0.027723
2,SVC_rbf,0.836727,0.029162
3,SVC_sigmoid,0.719700,0.057059
4,LogReg,0.949871,0.013335


Se puede observar que LogisticRegression fue el modelo con mejor desempeño, al presentar la media más alta del AUC (0.949871) y la menor desviación estándar. Esto indica que no solo fue el modelo más preciso en promedio, sino también el más estable. Por otro lado, los modelos SVC con kernels linear, poly y rbf obtuvieron resultados inferiores, siendo mucho más notable la alta desviación del kernel linear. 

# Addendum

Métricos disponibles para clasificación:
- ‘accuracy’
- ‘balanced_accuracy’
- ‘top_k_accuracy’
- ‘average_precision’
- ‘neg_brier_score’
- ‘f1’
- ‘f1_micro’
- ‘f1_macro’
- ‘f1_weighted’
- ‘f1_samples’
- ‘neg_log_loss’
- ‘precision’ etc.
- ‘recall’ etc.
- ‘jaccard’ etc.
- ‘roc_auc’
- ‘roc_auc_ovr’
- ‘roc_auc_ovo’
- ‘roc_auc_ovr_weighted’
- ‘roc_auc_ovo_weighted’
- ‘d2_log_loss_score’

# References

[1] Shigeo Abe.Support Vector Machines for Pattern Classification,2Ed.Springer-Verlag London,2010. ISBN978-1-84996-097-7. URLhttps://www.springer.com/gp/book/9781849960977.

[2] Johan A K Suykens, Tony Van Gestel, Jos De Brabanter, BartDe Moor, and Joos Vandewalle.Least Squares Support VectorMachines. World Scientific,2002. ISBN9789812381514. URLhttps://www.worldscientific.com/worldscibooks/10.1142/5089.

[3] Bradley, A. P. (1997). The use of the area under the ROC curve in the evaluation of machine learning algorithms. Pattern recognition, 30(7), 1145-1159. URL https://www.researchgate.net/post/how_can_I_interpret_the_ROC_curve_result